pipelime/piper/model.py

```diff
@@ -12,21 +13,96 @@ def pipelime_command(func=None, **kwargs):
+                    name=p.name, kind=p.kind, default=default, annotation=p.annotation
+                )
+
+            ann = p.annotation
+            if p.kind is p.VAR_POSITIONAL:
+                varpos_args = f"{p}, ".replace("NoneType", "None")
+                varpos_name = p.name
+                ann = (
+                    t.Sequence
+                    if p.annotation is inspect.Signature.empty
+                    else t.Sequence[p.annotation]
+                )
+            elif p.kind is p.POSITIONAL_ONLY:
+                posonly_args += f"{p}, ".replace("NoneType", "None")
+                posonly_names.append(p.name)
+            elif p.kind is p.POSITIONAL_OR_KEYWORD:
+                poskw_args += f"{p}, ".replace("NoneType", "None")
+                poskw_names.append(p.name)
+
+            if ann is inspect.Signature.empty:
                 return (t.Any, value) if value is ... else value
-            return (p.annotation, value)
+            return (ann, value)
 
         fsig = inspect.signature(func)
-        fields = {n: _make_field(p) for n, p in fsig.parameters.items()}
+        fields = {
+            n: _make_field(p)
+            for n, p in fsig.parameters.items()
+            if p.kind is not p.VAR_KEYWORD
+        }
+
+        if any(p.kind is p.VAR_KEYWORD for p in fsig.parameters.values()):
+            kwargs.setdefault("extra", "allow")
+            add_data_arg = True
+        else:
+            add_data_arg = any(
+                p.kind in (p.POSITIONAL_OR_KEYWORD, p.KEYWORD_ONLY)
+                for p in fsig.parameters.values()
+            )
 
-        if "title" not in kwargs:
-            kwargs["title"] = func.__name__
+        kwargs.setdefault("title", func.__name__)
+
+        if posonly_args:
+            posonly_args += "/, "
+        pos_names = posonly_names + poskw_names
 
         class _FnModel(PipelimeCommand):
             def run(self):
-                func(**self.dict())
+                func(
+                    *[getattr(self, n) for n in pos_names],
+                    *getattr(self, varpos_name, tuple()),
+                    **self.dict(exclude=set(pos_names + [varpos_name])),
+                )
+
+        local_scope = {**globals(), **locals()}
+        _exfn = """
+from typing import *
+def initfn(self, {}{}{}{}
+""".format(
+            posonly_args,
+            poskw_args,
+            varpos_args,
+            "**__data):\n" if add_data_arg else "):\n    __data = {}",
+        )
+
+        if varpos_name:
+            _exfn += "    __data.setdefault('{0}', {0})\n".format(varpos_name)
+        for p in pos_names:
+            _exfn += "    __data.setdefault('{0}', {0})\n".format(p)
+        _exfn += "    super(_FnModel, self).__init__(**__data)\n"
+
+        exec(_exfn, local_scope)
+        _FnModel.__init__ = local_scope["initfn"]
 
         fmodel = create_model(
             func.__name__.replace("_", " ").capitalize().strip() + "Command",
```

Comment:
La pipelime_command è molto complessa e difficile da seguire, vale la pena splittare in più parti. 

La parte finale della funzione fa code generation, siamo sicuri non ci sia un modo più semplice di creare quel model?

Reply:
- La code generation è necessaria perché deve essere creato un modello pydantic PipelimeCommand i cui field sono estratti dinamicamente dalla signature della funzione. Ti viene in mente un modo alternativo?
- La lunghezza di `pipelime_command` è 3 righe, la complessità deriva dal fatto di avere il decoratore vero e proprio `_make_cmd` innestato, come è prassi per nasconderlo. A sua volta `_make_cmd` usa la funzione innestata `_make_field` per incapsulare meglio la generazione dinamica dei fields. Di nuovo, la funzione è innestata per evidenziarne l'uso privato, cioè è solo per strutturare meglio il codice interno a `_make_cmd`, non dovrebbe essere vista dall'esterno. Tirando fuori queste due funzioni a livello visivo diventa tutto più "corto", ma a livello logico ci vedo delle falle. Intanto aggiungo commenti.